app/script/conversation/ClientMismatchHandler.js

```diff
@@ -0,0 +1,208 @@
+    };
+
+    const _removeDeletedUser = userId => {
+      if (payload && !Object.keys(payload.recipients[userId]).length) {
+        delete payload.recipients[userId];
+      }
+    };
+
+    return Promise.all(this._mapRecipients(recipients, _removeDeletedClient, _removeDeletedUser)).then(() => {
+      this.conversationRepository.verification_state_handler.on_client_removed(Object.keys(recipients));
+      return payload;
+    });
+  }
+
+  /**
+   * Handle the missing client mismatch.
+   *
+   * @private
+   * @param {Object} recipients - User client map containing redundant clients
+   * @param {Object} payload - Optional payload of the failed request
+   * @param {z.proto.GenericMessage} genericMessage - Protobuffer message to be sent
+   * @returns {Promise} Resolves with the updated payload
+   */
+  _handleClientMismatchMissing(recipients, payload, genericMessage) {
+    if (!payload || _.isEmpty(recipients)) {
+      return Promise.resolve(payload);
+    }
+    this.logger.debug(`Message is missing clients of '${Object.keys(recipients).length}' users`, recipients);
+
+    return this.cryptographyRepository
+      .encrypt_generic_message(recipients, genericMessage, payload)
+      .then(updatedPayload => {
+        payload = updatedPayload;
+
+        const _addMissingClient = (userId, clientId) => {
+          return this.userRepository.add_client_to_user(userId, new z.client.Client({id: clientId}));
+        };
+
+        return Promise.all(this._mapRecipients(recipients, _addMissingClient));
+      })
+      .then(() => {
+        this.conversationRepository.verification_state_handler.on_client_add(Object.keys(recipients));
+        return payload;
+      });
+  }
+
+  /**
+   * Handle the redundant client mismatch.
+
+   * @note Contains clients of which the backend is sure that they should not be recipient of a message but cannot say whether they exist.
+   *   Normally only contains clients of users no longer participating in a conversation.
+   *   Sometimes clients of the self user are listed. Thus we cannot remove the payload for all the clients of a user without checking.
+   * @private
+   *
+   * @param {Object} recipients - User client map containing redundant clients
+   * @param {Object} payload - Optional payload of the failed request
+   * @param {string} conversationId - ID of conversation the message was sent in
+   * @returns {Promise} Resolves with the updated payload
+   */
+  _handleClientMismatchRedundant(recipients, payload, conversationId) {
+    if (_.isEmpty(recipients)) {
+      return Promise.resolve(payload);
+    }
+    this.logger.debug(`Message contains redundant clients of '${Object.keys(recipients).length}' users`, recipients);
+
+    return this.conversationRepository
+      .get_conversation_by_id(conversationId)
+      .catch(error => {
+        if (error.type !== z.conversation.ConversationError.TYPE.NOT_FOUND) {
+          throw error;
+        }
+      })
+      .then(conversationEt => {
```

Comment:
I suggest to skip the `conversation_et` and `conversation_ets` pattern: `conversationEt` -> `conversationEntity`